engine/engine_test.go

```diff
@@ -0,0 +1,268 @@
+	game := New()
+	cardIndex := 3
+
+	game.SetPlayer(0, 0, maxHandSize, cloneHand(handOne))
+	game.SetPlayer(1, 0, maxHandSize, cloneHand(handTwo))
+	game.SetPlayer(2, 0, maxHandSize, cloneHand(handThree))
+	game.SetPlayer(3, 0, maxHandSize, cloneHand(handFour))
+
+	player, _ := game.Player(0)
+
+	for _, card := range player.hand {
+		if card.exposed || card.played {
+			t.Error("Cards should be dealt unplayed and unexposed")
+		}
+	}
+
+	err := game.PlayUp(0, handTwo[cardIndex+1])
+
+	if err == nil {
+		t.Error("should only be able to play a card in the player's hand")
+	}
+
+	err = game.PlayUp(0, handOne[cardIndex])
+
+	if err != nil {
+		t.Error("should be able to play a card in player's hand")
+	}
+
+	for i, card := range player.hand {
+		if i == cardIndex && (!card.played || !card.exposed) {
+			t.Error("cards should be be playable")
+		} else if i != cardIndex && (card.played || card.exposed) {
+			t.Error("only the selected card should be played")
+		}
+	}
+
+	err = game.PlayUp(0, handOne[cardIndex])
+
+	if err == nil {
+		t.Error("should not be able to play the same card twice")
+	}
+}
+
+func TestViewAs(t *testing.T) {
+	game := New()
+
+	game.SetPlayer(0, 0, maxHandSize, cloneHand(handOne))
+	game.SetPlayer(1, 0, maxHandSize, cloneHand(handTwo))
+	game.SetPlayer(2, 0, maxHandSize, cloneHand(handThree))
+	game.SetPlayer(3, 0, maxHandSize, cloneHand(handFour))
+
+	view := game.ViewAs(playerThree)
+
+	err := view.PlayUp(2, handThree[3])
+
+	if err == nil {
+		t.Error("should not be able to play cards in a readonly state")
+	}
+
+	err = view.SetPlayer(0, 10, 12, handOne)
+
+	if err == nil {
+		t.Error("should not be able to set players in a readonly state")
+	}
+
+	for i, player := range view.players {
+		if player.cardCount != maxHandSize {
+			t.Errorf("player %d should have %d cards", i, maxHandSize)
+		}
+
+		if i == playerThree {
+			err := compareHand(player.hand, handThree)
+
+			if err != nil {
+				t.Errorf("expected:\n\n%v\nfound:%v\n", handThree, player.hand)
+			}
+
+		} else if len(player.hand) != 0 {
+			t.Errorf("oppenent hands should not be viewable. saw %v", player.hand)
+		}
+	}
+
+	playedCard := handFour[5]
+
+	game.PlayUp(playerFour, playedCard)
+	view = game.ViewAs(playerThree)
+
+	if len(view.players[playerFour].hand) != 1 {
+		t.Error("player should only have one visible card")
+	}
+
+	actual := view.players[playerFour].hand[0]
+
+	if err := compareCard(actual, playedCard); err != nil {
+		t.Error("the wrong card is showing")
+	}
+}
+
+// PRIVATE HELPER FUNCTIONS
+
+func compareHand(actual []Card, expected []Card) error {
+	// If the lengths don't match, they must be different.
+	if len(actual) != len(expected) {
+		return fmt.Errorf("expected:\n%v\nfound:\n%v", expected, actual)
+	}
+
+	// If each card matches, and the length is the same, they must match.
+	for i, card := range actual {
+		return compareCard(card, expected[i])
+	}
+
+	return nil
+}
+
+func compareCard(actual Card, expected Card) error {
```

Comment:
Should we eventually add an equals methods to card like this?

Reply:
Yes. I didn't even think of that!

Reply:
I'm going to defer this for now, but we are going to need comparison methods later when we are actually programming the game rules.